daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -1178,24 +1215,32 @@ private[lf] object Speedy {
     * evaluated. If 'KCatch' is encountered naturally, then 'fin' is evaluated.
     */
   private[speedy] final case class KCatch(
-      savedBase: Int,
+      machine: Machine,
       handler: SExpr,
       fin: SExpr,
-      frame: Frame,
-      actuals: Actuals,
-      envSize: Int)
-      extends Kont
+  ) extends Kont
       with SomeArrayEquals {
-    def execute(v: SValue, machine: Machine) = {
-      machine.restoreBase(savedBase)
+
+    // We call [markBase] (as standard) so the continuation may access its temporaries.
+    // In addition [env.size] is recorded for use in [tryHandleException] to allow the
+    // env-stack to be unwound correctly when an exception is thrown.
```

Comment:
It is still unclear to me why the stack height returned by `makeBase` is not sufficient. After successfully leaving the `try { ... }` block or before entering the `catch { ... }` block, you want to restore to the same stack height, don't you?

Reply:
The main reason the code reverts to the env.size is because that's what it was doing before!

This would be necessary if the handler continuation requires access to the temps of the context when it was created. Which seems reasonable.

But I suspect that it never does use any temps, because try-catch is only used to support submitMustFail. i.e. try-catch is not directly exposed to the user

I will investigate some more.

Reply:
Aren't the contexts at which the handler was created and the context to which we want to jump back when no exception was raised exactly the same?

Reply:
Yes. But the stack height we want is `env.size`, not `envBase` . (`markBase` increases `envBase` to `env.size` and returns the old value of `envBase`, which doesn't include any temporaries on the stack between (old) `envBase` and `env.size`. )

The general form of the *try-catch* construct (`SECatch`) is: 
    `try BODY catch HANDLER finally FIN end-try` (made up syntax)
Upper case names are arbitrary expressions:
    `HANDLER` is evaluated if an exception is thrown during evaluation of `BODY`
    `FIN` is evaluated after `BODY` if no exception is thrown

If a try-catch expression is embedded in an enclosing `let` expression, any variables bound in this let-expression should be in scope for `BODY`, `HANDLER` and `FIN`, and will be temporaries on the env-stack.  The env-stack should only be popped back to `envBase` when evaluation leaves the context of `HANDLER` or `FIN`, either by returning a value, or tail calling. If an exception is thrown during the evaluation of `BODY`, we need to rewind the stack back to `env.size` before evaluation of `HANDER`.  Hence why we store `env.size` in the closure.


However....

As it turns out... both the `FIN` and the `HANDER` will not reference any variables, but will always be simple atomic values... we know this because the try-catch expression is only constructed by the speedy-compiler in one situation:

```
      labeledUnaryFunction("submitMustFail") { tokenPos =>
        let(SBSBeginCommit(optLoc)(compile(party), svar(tokenPos))) { _ =>
          let(SECatch(app(compile(update), svar(tokenPos)), SEValue.True, SEValue.False)) {
            resultPos =>
              SBSEndCommit(mustFail = true)(svar(resultPos), svar(tokenPos))
          }
        }
      }
```

so... `env.size` and the old/new values `envBase` will all be identical.
However, I think if we want to make use of this knowledge, we should probably de-generalize the `SECatch` class to represent only the limited case we actually require.

Reply:
I don't think there is any need to do anything here for this PR.

Reply:
I think I wrapped my head around this. When the `try ... catch ...` expression is in tail position, how deep we need to unwind the stack depends on whether the guarded expression was successful or failed. In the case of success, we need to pop the locals we had before entering the `try` block but in case of failure, we might (in general) need these locals in the exception handler. If the `try ... catch ...` block is `let` bound, then both stack heights will be the same. Does that explanation make sense? If so, could you please slightly extend the comment in that direction? Thank you.